graylog2-server/src/main/java/org/graylog2/inputs/InputDiagnosticService.java

```diff
@@ -0,0 +1,142 @@
+import org.graylog.plugins.views.search.Query;
+import org.graylog.plugins.views.search.QueryResult;
+import org.graylog.plugins.views.search.Search;
+import org.graylog.plugins.views.search.SearchJob;
+import org.graylog.plugins.views.search.SearchType;
+import org.graylog.plugins.views.search.elasticsearch.ElasticsearchQueryString;
+import org.graylog.plugins.views.search.engine.SearchExecutor;
+import org.graylog.plugins.views.search.errors.SearchError;
+import org.graylog.plugins.views.search.permissions.SearchUser;
+import org.graylog.plugins.views.search.rest.ExecutionState;
+import org.graylog.plugins.views.search.rest.SearchJobDTO;
+import org.graylog.plugins.views.search.searchtypes.pivot.Pivot;
+import org.graylog.plugins.views.search.searchtypes.pivot.PivotResult;
+import org.graylog.plugins.views.search.searchtypes.pivot.buckets.Values;
+import org.graylog.plugins.views.search.searchtypes.pivot.series.Count;
+import org.graylog2.database.NotFoundException;
+import org.graylog2.plugin.indexer.searches.timeranges.RelativeRange;
+import org.graylog2.rest.models.system.inputs.responses.InputDiagnostics;
+import org.graylog2.streams.StreamService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.graylog2.plugin.Message.FIELD_GL2_SOURCE_INPUT;
+import static org.graylog2.rest.models.system.inputs.responses.InputDiagnostics.EMPTY_DIAGNOSTICS;
+
+public class InputDiagnosticService {
+    private static final Logger LOG = LoggerFactory.getLogger(InputDiagnosticService.class);
+
+    private static final String QUERY_ID = "input_diagnostics_streams_query";
+    private static final String PIVOT_ID = "input_diagnostics_streams_pivot";
+
+    private final SearchExecutor searchExecutor;
+    private final StreamService streamService;
+
+    @Inject
+    public InputDiagnosticService(SearchExecutor searchExecutor,
+                                  StreamService streamService) {
+        this.searchExecutor = searchExecutor;
+        this.streamService = streamService;
+    }
+
+    public InputDiagnostics getInputDiagnostics(
+            Input input, SearchUser searchUser) {
+        final Search search = buildSearch(input);
+        final SearchJob searchJob = searchExecutor.executeSync(search, searchUser, ExecutionState.empty());
+        final SearchJobDTO searchJobDTO = SearchJobDTO.fromSearchJob(searchJob);
+        final QueryResult queryResult = searchJobDTO.results().get(QUERY_ID);
+
+        final Set<SearchError> errors = queryResult.errors();
+        if (errors != null && !errors.isEmpty()) {
+            LOG.error("An error occurred while executing aggregation: {}",
+                    errors.stream().map(SearchError::description).collect(Collectors.joining(", ")));
+            return EMPTY_DIAGNOSTICS;
```

Comment:
It may be better to pass the error message to the UI here, otherwise it would be a bit confusing for the user